### PR TITLE
refactor: use `java.net.URI` for parsing URLs in `UrlUtils`

### PR DESCRIPTION
**Pull request by lcian**
## :scroll: Description

This changes the internal logic used for parsing URLs in `UrlUtils` from regex based to using `java.net.URI`.
This will allow the SDK to handle the cases described in https://github.com/getsentry/sentry-java/issues/2690 and https://github.com/getsentry/team-sdks/issues/39.
`java.net.URI` is known to still not be fully compliant with RFCs 3986 and 3987 (https://cr.openjdk.org/%7Edfuchs/writeups/updating-uri/) but this is still an improvement with respect to the cases described in the issues linked above, and could also ensure correct handling for other corner cases that we are not aware of.
We still use `java.net.URL` but just for validating that an URI is a valid URL.

This could affect grouping for existing issues that use (parts of) an URL as the message when the URL falls in the affected corner cases.

## :bulb: Motivation and Context

Closes https://github.com/getsentry/sentry-java/issues/2690

## :green_heart: How did you test it?
Added some tests and changed existing ones

## :pencil: Checklist

- [X] I added tests to verify the changes.
- [X] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] I updated the wizard if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [X] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

**Review by adinauer — APPROVED** [review]
Mostly LGTM

File: sentry/src/main/java/io/sentry/util/UrlUtils.java [hunk 1]
```diff
@@ -15,123 +12,55 @@
...
-      final int fragmentSeparatorIndex = url.indexOf("#");
-      if (fragmentSeparatorIndex >= 0) {
-        return url.substring(0, fragmentSeparatorIndex).trim();
-      } else {
-        return url;
-      }
+    if (url.startsWith("@")) {
+      return SENSITIVE_DATA_SUBSTITUTE + url;
     }
+    final @NotNull String userInfo = url.substring(0, url.indexOf('@') - 1);
```
[the remark this answers is not in the document]
Comment (lcian) [reply sentry/src/main/java/io/sentry/util/UrlUtils.java:58]: good catch!
